src/librustc_typeck/check/cast.rs

```diff
@@ -536,7 +536,10 @@ impl<'a, 'tcx> CastCheck<'tcx> {
                 match self.expr_ty.kind {
                     ty::FnDef(..) => {
                         // Attempt a coercion to a fn pointer type.
-                        let f = self.expr_ty.fn_sig(fcx.tcx);
+                        let f = fcx.normalize_associated_types_in(
+                            self.expr.span,
+                            &self.expr_ty.fn_sig(fcx.tcx),
+                        );
```

Comment:
cc @nikomatsakis This makes sense to me.

Reply:
Yes, it seems like it is necessary, given that `fn_sig` returns the result of data from the `def_id` ..